### PR TITLE
v3 checkpoint PR

### PR DESCRIPTION
**Pull request by usergenic**
A checkpoint for v3 work.  Inlining HTML support, converted all tests to the promise style, added source maps support for tests, and other minor fixes.  (27 tests passing, 21 failing, 7 pending)

### CODE REVIEW

File: src/bundler.ts [hunk 1]
```diff
@@ -125,12 +125,41 @@ class Bundler {
...
     const hidden = dom5.constructors.element('div');
     dom5.setAttribute(hidden, 'hidden', '');
     dom5.setAttribute(hidden, 'by-vulcanize', '');
-    this.removeElementAndNewline(node, hidden);
-    dom5.append(hidden, node);
+    return hidden;
+  }
+
+  /**
+   * Replace htmlImport
```
Comment (garlicnation) [inline src/bundler.ts]: Fix this comment.

File: src/bundler.ts [hunk 2]
```diff
@@ -125,12 +125,41 @@ class Bundler {
...
+    const resolved = url.resolve(documentUrl, rawUrl);
+    const backingScript: ScannedDocument =
+        await analyzer._analyzeResolved(resolved);
+    const scriptContent = backingScript.document.contents;
+    dom5.removeAttribute(externalScript, 'src');
+    dom5.setTextContent(externalScript, scriptContent);
+  }
+
+  /**
+   * Replace htmlImport
```
Comment (garlicnation) [inline src/bundler.ts]: Fix this comment.

File: src/bundler.ts [hunk 3]
```diff
@@ -150,31 +179,64 @@ class Bundler {
         await analyzer._analyzeResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent;
+    if (matchers.afterHiddenDiv(htmlImport)) {
```
Comment (garlicnation) [inline src/bundler.ts]: This function feels too long. I'd like to see the hidden div logic split out to make it clearer what is pure inlining vs our hacks.

Also, splitting out the hidden div functionality should enable some highly needed unit tests around it.

File: src/bundler.ts [hunk 4]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
     const body = dom5.query(newDocument, matchers.body);
-    const head = dom5.query(newDocument, matchers.head);
+    const hiddenDiv = this.getHiddenNode();
+
     const getNextImport = () => dom5.query(newDocument, matchers.htmlImport);
     const elementInHead = dom5.predicates.parentMatches(matchers.head);
     const inlinedImports = new Set<string>();
-    let c = 1;
+
     for (let nextImport; nextImport = getNextImport();) {
```
Comment (garlicnation) [inline src/bundler.ts]: If you change this to a while loop, we can get tslint to stop complaining.

File: src/matchers.ts [hunk 5]
```diff
@@ -16,46 +16,73 @@
...
+
+export const hiddenDiv = predicates.AND(
+    predicates.hasTagName('div'), predicates.hasAttr('hidden'),
+    predicates.hasAttr('by-vulcanize'));
+
+export const inHiddenDiv = predicates.parentMatches(hiddenDiv);
+const nextToHiddenDiv = (offset) => {
+  return (node) => {
+    const siblings = node.parentNode.childNodes;
+    const hiddenDivIndex = siblings.indexOf(node) + offset;
```
Comment (garlicnation) [inline src/matchers.ts]: This matcher needs to deal with a number of edge cases. Whitespace-only text nodes should be ignored, text nodes with meaningful space should be preserved. Comments should be ignored if --strip-comments is set. License comments should be deduplicated and moved to the start of the document.

File: src/test/test.ts [hunk 6]
```diff
@@ -253,309 +255,277 @@ suite('Vulcan', function() {
...
-      process(inputPath)
+      bundle(inputPath)
           .then((doc) => {
             assert.equal(dom5.queryAll(doc, imports).length, 0);
             done();
           })
-          .catch((error) => {
-            done(error);
-          });
+          .catch(done);
```
Comment (garlicnation) [inline src/test/test.ts]: Does this successfully report errors? I could have sworn I tried this pattern and it didn't work, but it might have had to do with an incomplete refactor of `bundle(`

Comment (usergenic) [inline src/test/test.ts]: Yeah, there's no magic.  Verified that output is equivalent to write .catch((e) => done(e));

File: src/bundler.ts (hunk 3 above)
Comment (usergenic) [inline src/bundler.ts]: putting a TODO here for inlineImportHtml refactoring branch.

File: src/ast-utils.ts [hunk 7]
```diff
@@ -24,7 +24,7 @@ export default class ASTUtils {
     }
   }
 
-  static prependMultiple(target, nodes: ASTNode[]) {
+  private static prependMultiple(target, nodes: ASTNode[]) {
```
Comment (justinfagnani) [inline src/ast-utils.ts]: style: all private names should start with `_` so that it's apparent that it's private-ish after compilation.

File: src/ast-utils.ts [hunk 8]
```diff
@@ -34,6 +34,15 @@ export default class ASTUtils {
     }
   }
 
+  static insertAllBefore(target, oldNode: ASTNode, nodes: ASTNode[]) {
```
Comment (justinfagnani) [inline src/ast-utils.ts]: why is this a static method? Can these just be exported functions?

File: src/bundler.ts [hunk 9]
```diff
@@ -125,12 +125,41 @@ class Bundler {
     return false;
   }
 
-  hide(node) {
+  getHiddenNode(): ASTNode {
```
Comment (justinfagnani) [inline src/bundler.ts]: comment?

File: src/ast-utils.ts [hunk 10]
```diff
@@ -34,6 +34,15 @@ export default class ASTUtils {
...
+    let lastNode = oldNode;
+    for (let n = nodes.length - 1; n >= 0; n--) {
+      let node = nodes[n];
```
Comment (justinfagnani) [inline src/ast-utils.ts]: const

File: src/bundler.ts [hunk 11]
```diff
@@ -150,31 +179,64 @@ class Bundler {
         await analyzer._analyzeResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent;
```
Comment (justinfagnani) [inline src/bundler.ts]: add type annotation

File: src/bundler.ts (hunk 3 above)
Comment (justinfagnani) [inline src/bundler.ts]: +1

File: src/bundler.ts [hunk 12]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
+      importParent = dom5.nodeWalkPrior(htmlImport, matchers.hiddenDiv);
+      dom5.remove(htmlImport);
+      dom5.append(importParent, htmlImport);
+    } else if (matchers.beforeHiddenDiv(htmlImport)) {
+      const index = htmlImport.parentNode.childNodes.indexOf(htmlImport);
+      importParent = htmlImport.parentNode.childNodes[index + 1];
+      dom5.remove(htmlImport);
+      ASTUtils.prepend(importParent, htmlImport);
+    } else if (!matchers.inHiddenDiv(htmlImport)) {
+      let hiddenDiv = this.getHiddenNode();
```
Comment (justinfagnani) [inline src/bundler.ts]: const

File: src/bundler.ts [hunk 13]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
+    } else {
+      importParent = htmlImport.parentNode;
+    }
+
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
   async bundle(url: string): Promise<ASTNode> {
```
Comment (justinfagnani) [inline src/bundler.ts]: add comment

File: src/bundler.ts [hunk 14]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
     const analyzer: Analyzer = new Analyzer(this.opts);
-    const analyzed: Document = await analyzer.analyzeRoot(url);
-    const newDocument = dom5.parse(analyzed.parsedDocument.contents);
+    const analyzedRoot: Document = await analyzer.analyzeRoot(url);
+    const newDocument = dom5.parse(analyzedRoot.parsedDocument.contents);
```
Comment (justinfagnani) [inline src/bundler.ts]: did you try cloning the already parsed AST?

File: src/bundler.ts [hunk 15]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
   async bundle(url: string): Promise<ASTNode> {
     const analyzer: Analyzer = new Analyzer(this.opts);
-    const analyzed: Document = await analyzer.analyzeRoot(url);
-    const newDocument = dom5.parse(analyzed.parsedDocument.contents);
+    const analyzedRoot: Document = await analyzer.analyzeRoot(url);
```
Comment (justinfagnani) [inline src/bundler.ts]: you shouldn't need the type annotations on these consts

File: src/bundler.ts (hunk 4 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/bundler.ts [hunk 16]
```diff
@@ -150,31 +179,64 @@ class Bundler {
...
-    let c = 1;
+
     for (let nextImport; nextImport = getNextImport();) {
       // If the import is in head, move all subsequent nodes to body.
       if (elementInHead(nextImport)) {
+        // Put the hiddenDiv in the body the first time we need it.
+        if (!hiddenDiv.parentNode) {
+          ASTUtils.prepend(body, hiddenDiv);
+        }
         // This function needs a better name.
```
Comment (justinfagnani) [inline src/bundler.ts]: convert to a TODO

File: src/matchers.ts (hunk 5 above)
Comment (usergenic) [inline src/matchers.ts]: I'm working off the assumption that tests exist for edge cases you mentioned.  Adding TODO to ensure we have have them.

File: src/test/test.ts (hunk 6 above)
Comment (justinfagnani) [inline src/test/test.ts]: You should remove the done parameter and return the promise.

File: src/test/test.ts [hunk 17]
```diff
@@ -253,309 +255,277 @@ suite('Vulcan', function() {
...
-        if (err) {
-          return done(err);
-        }
-        const svg = dom5.query(doc, preds.hasTagName('svg'));
-        assert.equal(svg.childNodes.filter(dom5.isElement).length, 6);
-        done();
-      });
-    });
+  test('svg is nested correctly', function(done) {
+    bundle(inputPath)
```
Comment (justinfagnani) [inline src/test/test.ts]: Remove done argument and return the promise

File: src/bundler.ts (hunk 2 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/bundler.ts (hunk 1 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/ast-utils.ts (hunk 8 above)
Comment (usergenic) [inline src/ast-utils.ts]: SGTM.  I'll change it.

File: src/ast-utils.ts (hunk 10 above)
Comment (usergenic) [inline src/ast-utils.ts]: Done.

File: src/bundler.ts (hunk 11 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/bundler.ts (hunk 9 above)
Comment (usergenic) [inline src/bundler.ts]: Done. Added a TODO for renaming method as well to something more intention-revealing.

File: src/ast-utils.ts (hunk 8 above)
Comment (usergenic) [inline src/ast-utils.ts]: Done.

File: src/test/test.ts (hunk 6 above)
Comment (usergenic) [inline src/test/test.ts]: :derp:

Well that's like a million times better.

File: src/bundler.ts [hunk 18]
```diff
@@ -147,33 +183,69 @@ class Bundler {
...
+      importParent = htmlImport.parentNode;
+    }
+
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
   async bundle(url: string): Promise<ASTNode> {
     const analyzer: Analyzer = new Analyzer(this.opts);
```
Comment (justinfagnani) [inline src/bundler.ts]: We want the Bundler to take an Analyzer, probably in the constructor, rather than construct it's own.

File: src/bundler.ts [hunk 19]
```diff
@@ -30,7 +30,7 @@ import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
 import constants from './constants';
 import * as matchers from './matchers';
 import PathResolver from './pathresolver';
-import ASTUtils from './ast-utils';
+import * as ASTUtils from './ast-utils';
 
 
 function buildLoader(config: any) {
```
Comment (justinfagnani) [inline src/bundler.ts]: can this be removed?

File: src/bundler.ts [hunk 20]
```diff
@@ -55,7 +55,7 @@ class Bundler {
                     String(opts.abspath).trim() !== '') ?
         path.resolve(opts.abspath) :
         null;
-    this.pathResolver = new PathResolver(this.abspath);
+    this.pathResolver = new PathResolver(Boolean(this.abspath));
```
Comment (justinfagnani) [inline src/bundler.ts]: is pathResolver still used?

File: src/ast-utils.ts [hunk 21]
```diff
@@ -15,32 +15,39 @@
...
 import {ASTNode} from 'parse5';
 
-export default class ASTUtils {
-  static prepend(parent, node) {
-    if (parent.childNodes && parent.childNodes.length) {
-      dom5.insertBefore(parent, parent.childNodes[0], node);
-    } else {
-      dom5.append(parent, node);
-    }
+export function prepend(parent, node) {
```
Comment (justinfagnani) [inline src/ast-utils.ts]: Add type annotation to parameters.

Comment (justinfagnani) [inline src/ast-utils.ts]: Is `noImplcitAny` set to `true` in tsconfig?

File: src/ast-utils.ts [hunk 22]
```diff
@@ -15,32 +15,39 @@
...
 
-  static prependMultiple(target, nodes: ASTNode[]) {
-    let moveIndex = nodes.length - 1;
-    while (moveIndex >= 0) {
-      const nodeToMove = nodes[moveIndex];
-      dom5.remove(nodeToMove);
-      ASTUtils.prepend(target, nodeToMove);
-      moveIndex--;
-    }
+function _prependMultiple(target, nodes: ASTNode[]) {
```
Comment (justinfagnani) [inline src/ast-utils.ts]: type annotation for `target` here and in `insertAllBefore`

File: src/bundler.ts [hunk 23]
```diff
@@ -125,18 +125,54 @@ class Bundler {
...
     const hidden = dom5.constructors.element('div');
     dom5.setAttribute(hidden, 'hidden', '');
     dom5.setAttribute(hidden, 'by-vulcanize', '');
-    this.removeElementAndNewline(node, hidden);
-    dom5.append(hidden, node);
+    return hidden;
+  }
+
+  /**
+   * Inline external scripts <script src="*">
```
Comment (justinfagnani) [inline src/bundler.ts]: Where is this script inlined? From the parameter names, I'm having trouble which ones might be mutated.

File: src/bundler.ts [hunk 24]
```diff
@@ -125,18 +125,54 @@ class Bundler {
...
+   */
+  async inlineScript(
+      documentUrl: string, externalScript: ASTNode,
+      analyzer: Analyzer): Promise<void> {
```
Comment (justinfagnani) [inline src/bundler.ts]: If Bundler owns an Analyzer instance, it would need to be passed in here.

File: src/bundler.ts [hunk 25]
```diff
@@ -147,33 +183,69 @@ class Bundler {
     }
     inlined.add(resolved);
     const backingDocument: ScannedDocument =
-        await analyzer._analyzeResolved(resolved);
+        await analyzer._scanResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent: ASTNode;
+    if (matchers.afterHiddenDiv(htmlImport)) {
```
Comment (justinfagnani) [inline src/bundler.ts]: Can you add some comments to explain the significance of checking of the import is before or after a hidden div? Maybe break this out into a `getImportParent` method too.

File: src/bundler.ts (hunk 13 above)
Comment (justinfagnani) [inline src/bundler.ts]: I'm curious how `bundle()` relates to the options given to the constructor. Is it expected that `bundle` will only be called on one particular URL, since the constructor has seemlingly document specific things, like excludes?

File: src/bundler.ts [hunk 26]
```diff
@@ -147,33 +183,69 @@ class Bundler {
...
+    const newDocument = dom5.parse(analyzedRoot.parsedDocument.contents);
+
+    // Create a hidden div to target.
     const body = dom5.query(newDocument, matchers.body);
-    const head = dom5.query(newDocument, matchers.head);
-    const getNextImport = () => dom5.query(newDocument, matchers.htmlImport);
+    const hiddenDiv = this.getHiddenNode();
+
+    const getNextHtmlImport = () =>
+        dom5.query(newDocument, matchers.htmlImport);
```
Comment (justinfagnani) [inline src/bundler.ts]: The scanned document should already have the list of imports, though I think we just removed the reference from Feature->Node. @rictic we may want that back.

File: src/bundler.ts [hunk 27]
```diff
@@ -147,33 +183,69 @@ class Bundler {
...
     const elementInHead = dom5.predicates.parentMatches(matchers.head);
-    const inlinedImports = new Set<string>();
-    let c = 1;
-    for (let nextImport; nextImport = getNextImport();) {
+    const inlinedHtmlImports = new Set<string>();
+
+    let nextHtmlImport;
+    while (nextHtmlImport = getNextHtmlImport()) {
```
Comment (justinfagnani) [inline src/bundler.ts]: It seems like there are potentially multiple walks over the document tree. Once here, and once for script inlining. Plus there's a re-parser of the contents. I wonder if this would be better as a walk over the original AST, cloning nodes to build up a new document on the fly...

File: src/test/test.ts (hunk 6 above)
Comment (justinfagnani) [inline src/test/test.ts]: 🎉

File: src/bundler.ts (hunk 26 above)
Comment (rictic) [inline src/bundler.ts]: Arg. So the main reason I was so happy about getting rid of it is that it makes serializing any features problematic. Particularly parse5's html nodes, as they have parent references.

Comment (usergenic) [inline src/bundler.ts]: @rictic when you say getting rid of "it" are you referring to the hidden div?

File: src/bundler.ts (hunk 16 above)
Comment (usergenic) [inline src/bundler.ts]: Done

File: src/test/test.ts (hunk 17 above)
Comment (usergenic) [inline src/test/test.ts]: All "done" args removed from tests.  All tests return promise now.

File: src/ast-utils.ts (hunk 7 above)
Comment (usergenic) [inline src/ast-utils.ts]: _Done.

File: src/bundler.ts (hunk 12 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/bundler.ts (hunk 13 above)
Comment (usergenic) [inline src/bundler.ts]: I've documented how I expect bundle will be called.  Bundler's configuration does appear to have lots of context-specific settings such as excludes etc.  Perhaps these will migrate into bundling options instead of Bundler options.  Will revisit this choice once tests passing.

File: src/bundler.ts (hunk 14 above)
Comment (usergenic) [inline src/bundler.ts]: Added a TODO above the newDocument assignment to look at either cloning or building it node-by-node.

File: src/bundler.ts (hunk 15 above)
Comment (usergenic) [inline src/bundler.ts]: Done

File: src/bundler.ts (hunk 27 above)
Comment (usergenic) [inline src/bundler.ts]: I put that suggestion in a TODO.  I think building up newDocument node by node is compelling.

File: src/ast-utils.ts (hunk 21 above)
Comment (usergenic) [inline src/ast-utils.ts]: noImplicitAny was set to false so I set it to true and also strictNullChecks to true.  And then I type'd all the things.

File: src/ast-utils.ts [hunk 28]
```diff
@@ -15,32 +15,40 @@
...
+  let lastNode = oldNode;
+  for (let n = nodes.length - 1; n >= 0; n--) {
+    const node = nodes[n];
+    dom5.insertBefore(target, lastNode, node);
+    lastNode = node;
   }
 }
+
+/**
+ * Move node and its subsequent siblings to target.
```
Comment (justinfagnani) [inline src/ast-utils.ts]: where in the target does it move them? it the target a container, or are the nodes moved to be right after it?

File: src/bundler.ts [hunk 29]
```diff
@@ -30,32 +30,35 @@ import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
...
+  redirects?: string[];
+  stripComments?: boolean;
+  stripExcludes?: string[];
 }
 
+
 class Bundler {
-  constructor(opts: any) {
+  constructor(opts: Options) {
+    this.analyzer = opts.analyzer!;
```
Comment (justinfagnani) [inline src/bundler.ts]: I'm not sure the `!` is correct here, as long as `analyzer` is optional in `Options`

File: src/bundler.ts [hunk 30]
```diff
@@ -30,32 +30,35 @@ import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
...
     // implicitStrip should be true by default
     this.implicitStrip =
         opts.implicitStrip === undefined ? true : Boolean(opts.implicitStrip);
     this.abspath = (String(opts.abspath) === opts.abspath &&
                     String(opts.abspath).trim() !== '') ?
         path.resolve(opts.abspath) :
-        null;
-    this.pathResolver = new PathResolver(this.abspath);
+        undefined;
```
Comment (justinfagnani) [inline src/bundler.ts]: Is this how clang is formatting a ternary? I'd rather see the `?` and ':' at the beginning of lines. Also, the `String()` boxing calls seem weird to me, isn't this just checking if the string is null or empty?

```
    this.abspath = (opts.abspath == null || opts.abspath.trim() == '')
      ? undefined
      : path.resolve(opts.abspath);
```

Also, also, `path.resolve(p)` implicitly depends on the CWD. I don't even know what abspath does, but if we're receiving a configured `Analyzer` is it necessary?

File: src/bundler.ts [hunk 31]
```diff
@@ -30,32 +30,35 @@ import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
...
-  let loader = new FSUrlLoader(root);
-  // TODO(garlicnation): Add noopResolver for external urls.
-  // TODO(garlicnation): Add redirectResolver for fakeprotocol:// urls
-  // TODO(garlicnation): Add noopResolver for excluded urls.
-  return loader;
+import * as ASTUtils from './ast-utils';
+
+
+export interface Options {
+  abspath?: string;
```
Comment (justinfagnani) [inline src/bundler.ts]: maybe not this second, but we should document these...

File: src/bundler.ts [hunk 32]
```diff
@@ -66,28 +69,23 @@ class Bundler {
...
-    this.fsResolver = opts.fsResolver;
     this.redirects = Array.isArray(opts.redirects) ? opts.redirects : [];
-    this.opts = {
-      urlLoader: new FSUrlLoader(opts.root || process.cwd()),
-    };
   }
-  implicitStrip: Boolean;
-  abspath;
+
+  abspath?: string;
```
Comment (justinfagnani) [inline src/bundler.ts]: style: put fields above the constructor

File: src/bundler.ts [hunk 33]
```diff
@@ -118,62 +116,138 @@ class Bundler {
...
+    }
+
+    ASTUtils.insertAllBefore(importParent, htmlImport, documentAst.childNodes!);
     dom5.remove(htmlImport);
-    ASTUtils.prependMultiple(importParent, documentAst.childNodes);
   }
 
+  /**
+   * Given a URL to an entry-point html document, produce a single document
+   * suitable for deployment.
```
Comment (justinfagnani) [inline src/bundler.ts]: "suitable for deployment" describes what the result might be used for, but not what the method does. "with HTML imports, external stylesheets and external scripts inlined, according to the options for this Bundler"?

File: src/bundler.ts [hunk 34]
```diff
@@ -118,62 +116,138 @@ class Bundler {
...
+    // in parsedDocument.  Try cloning or even building newDocument as analyzed
+    // document is walked.
+    const newDocument = dom5.parse(analyzedRoot.parsedDocument.contents);
+
+    // Create a hidden div to target.
+    const body: ASTNode = dom5.query(newDocument, matchers.body)!;
+    const hiddenDiv = this.getHiddenNode();
+
+    const getNextHtmlImport = () =>
+        dom5.query(newDocument, matchers.htmlImport);
```
Comment (justinfagnani) [inline src/bundler.ts]: this is going to query each time its called. Would it be correct to queryAll once and iterate through the list?

File: src/bundler.ts [hunk 35]
```diff
@@ -30,32 +30,35 @@ import {FSUrlLoader} from 'polymer-analyzer/lib/url-loader/fs-url-loader';
...
-  const excludes = config.excludes;
-  const fsResolver = config.fsResolver;
-  const redirects = config.redirects;
-  let root = abspath && path.resolve(abspath) || process.cwd();
-  let loader = new FSUrlLoader(root);
-  // TODO(garlicnation): Add noopResolver for external urls.
-  // TODO(garlicnation): Add redirectResolver for fakeprotocol:// urls
-  // TODO(garlicnation): Add noopResolver for excluded urls.
-  return loader;
+import * as ASTUtils from './ast-utils';
```
Comment (justinfagnani) [inline src/bundler.ts]: style: name the library lowercase so it doesn't look like static methods at call sites.

File: src/bundler.ts [hunk 36]
```diff
@@ -118,62 +116,138 @@ class Bundler {
...
     const backingDocument: ScannedDocument =
-        await analyzer._analyzeResolved(resolved);
+        await this.analyzer._scanResolved(resolved);
     const documentAst = dom5.parseFragment(backingDocument.document.contents);
     this.pathResolver.resolvePaths(documentAst, resolved, documentUrl);
-    const importParent = htmlImport.parentNode;
+    let importParent: ASTNode;
+    if (matchers.afterHiddenDiv(htmlImport)) {
+      importParent = dom5.nodeWalkPrior(htmlImport, matchers.hiddenDiv)!;
+      dom5.remove(htmlImport);
```
Comment (justinfagnani) [inline src/bundler.ts]: TODO: remove the `remove()` call when https://github.com/PolymerLabs/dom5/issues/35 is fixed

File: src/bundler.ts [hunk 37]
```diff
@@ -118,62 +116,138 @@ class Bundler {
...
+      dom5.append(importParent, htmlImport);
+    } else if (matchers.beforeHiddenDiv(htmlImport)) {
+      const index = htmlImport.parentNode!.childNodes!.indexOf(htmlImport);
+      importParent = htmlImport.parentNode!.childNodes![index + 1];
+      dom5.remove(htmlImport);
+      ASTUtils.prepend(importParent, htmlImport);
+    } else if (!matchers.inHiddenDiv(htmlImport)) {
+      const hiddenDiv = this.getHiddenNode();
+      dom5.replace(htmlImport, hiddenDiv);
+      dom5.append(hiddenDiv, htmlImport);
```
Comment (justinfagnani) [inline src/bundler.ts]: Is this correct? Don't you want to append the import's document contents?

File: src/pathresolver.ts [hunk 38]
```diff
@@ -28,21 +28,21 @@ class PathResolver {
   constructor(public abspath?: boolean) {
   }
 
-  isTemplatedUrl(href) {
+  isTemplatedUrl(href: string): boolean {
```
Comment (justinfagnani) [inline src/pathresolver.ts]: comment?

**Comment by justinfagnani** [comment]
@usergenic can you update .travis.yml to only run tests on node 4 and 6?

File: src/bundler.ts (hunk 30 above)
Comment (usergenic) [inline src/bundler.ts]: That whole segment of code was cargo'd in from the old vulcan.js.  It could be reworked to something with clearer intent.  I'll fix it up.  The String() bit does seem unnecessary.

The ternary formatting is clang's choice.

I'm not 100% sure what abspath is for.  I've not fully come to understand or encounter the way they are being exercised yet in tests.  I'll be documenting and/or removing each option fully in the Options type for bundler soon.

Comment (usergenic) [inline src/bundler.ts]: There is a big ugly PathResolver module cargo'd in from old vulcanize that probably needs to be refactored to take analyzer's resolver into account, but it apparently uses abspath.  Will replace or update it once I have all the tests covering path resolution working.

File: src/bundler.ts (hunk 31 above)
Comment (usergenic) [inline src/bundler.ts]: yeah definitely going to document them...once i know what they're for.

File: src/bundler.ts (hunk 32 above)
Comment (usergenic) [inline src/bundler.ts]: Done.

File: src/bundler.ts (hunk 33 above)
Comment (usergenic) [inline src/bundler.ts]: Agreed.  Thanks for writing that for me :)

File: src/bundler.ts (hunk 34 above)
Comment (usergenic) [inline src/bundler.ts]: That's a very good point.  I think perhaps this approach must have been a lazy way to ensure nested imports worked; definitely need to fix this.  Adding a TODO.

File: src/bundler.ts (hunk 35 above)
Comment (usergenic) [inline src/bundler.ts]: Ok.  Done.

File: src/bundler.ts (hunk 36 above)
Comment (usergenic) [inline src/bundler.ts]: Added.

File: src/pathresolver.ts (hunk 38 above)
Comment (usergenic) [inline src/pathresolver.ts]: PathResolver is going to undergo some refactoring.  I only added type annotations here to existing stuff because lint warnings.  Will add comments when refactored to support new root+abspath behavior.
